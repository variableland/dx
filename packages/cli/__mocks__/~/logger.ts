import type { AnyLogger } from "~/logger";

const StubLogger: AnyLogger = {
  info: () => {},
  error: () => {},
  debug: () => {},
  subdebug: () => () => {},
};

export { StubLogger as Log };
