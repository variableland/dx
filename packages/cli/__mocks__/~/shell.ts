import { vi } from "vitest";

export const $ = vi.fn(function fakeShell(strs: string[], ...args: any) {
  let output = "";
  let argsIndex = 0;

  const stringifyArg = (arg: any) => (typeof arg.stdout === "string" ? arg.stdout : arg);

  strs.forEach((str) => {
    if (str === "") {
      if (args[argsIndex]) {
        output += stringifyArg(args[argsIndex]);
        argsIndex += 1;
      }
    } else {
      output += str;
    }
  });

  return output;
});
