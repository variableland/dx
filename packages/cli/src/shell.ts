import { $ as make$ } from "zx";
import { quote } from "./utils/quote";
import { isRaw } from "./utils/is-raw";

export const $ = make$({
  verbose: true,
  quote: (arg: unknown) => {
    if (typeof arg === "string") {
      return quote(arg);
    }

    if (isRaw(arg)) {
      return arg.stdout;
    }

    throw TypeError(`Unsupported argument type: ${typeof arg}`);
  },
});
