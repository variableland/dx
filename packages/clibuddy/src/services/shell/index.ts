export * from "./create";
export * from "./shell";
export * from "./types";
export * from "./utils";
