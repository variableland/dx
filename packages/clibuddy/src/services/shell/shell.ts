import { type Options, type Shell, $ as make$ } from "zx";

type ShellOptions = Partial<Options>;

export class ShellService {
  #shell: Shell;
  #quiet: Shell;

  constructor(options: ShellOptions) {
    this.#shell = make$(options);
    this.#quiet = make$({ ...options, verbose: false });
  }

  get $() {
    return this.#shell;
  }

  get $quiet() {
    return this.#quiet;
  }
}
