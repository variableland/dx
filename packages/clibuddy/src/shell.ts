import fs from "node:fs";
import { $ as make$ } from "zx";

export const isRaw = (arg: unknown): arg is { stdout: string } =>
  typeof arg === "object" && arg !== null && "stdout" in arg && typeof arg.stdout === "string";

export const cwd = fs.realpathSync(process.env.PWD ?? process.cwd());

// Inspired by https://dub.sh/6tiHVgn
export function quote(arg: string) {
  if (/^[\w./:=@-]+$/i.test(arg) || arg === "") {
    return arg;
  }

  return arg
    .replace(/\\/g, "\\\\")
    .replace(/'/g, "\\'")
    .replace(/\f/g, "\\f")
    .replace(/\n/g, "\\n")
    .replace(/\r/g, "\\r")
    .replace(/\t/g, "\\t")
    .replace(/\v/g, "\\v")
    .replace(/\0/g, "\\0");
}

export const $ = make$({
  cwd,
  verbose: true,
  quote: (arg: unknown) => {
    if (typeof arg === "string") {
      return quote(arg);
    }

    if (isRaw(arg)) {
      return arg.stdout;
    }

    throw TypeError(`Unsupported argument type: ${typeof arg}`);
  },
});
