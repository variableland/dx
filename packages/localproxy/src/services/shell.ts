import { createShellService } from "@vlandoss/clibuddy";

export const quietShell = createShellService({
  quiet: true,
  verbose: false,
});

export const silentShell = quietShell.child({
  stdio: ["ignore", "ignore", "ignore"],
});

export const verboseShell = quietShell.child({
  quiet: false,
  verbose: true,
});
