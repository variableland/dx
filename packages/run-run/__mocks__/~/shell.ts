import { vi } from "vitest";
import { isRaw } from "../../src/utils/is-raw";

export const $ = vi.fn(function fakeShell(strs: string[], ...args: string[]) {
  let output = "";
  let argsIndex = 0;

  const stringifyArg = (arg: unknown) => (isRaw(arg) ? arg.stdout : arg);

  for (const str of strs) {
    if (str === "") {
      if (args[argsIndex]) {
        output += stringifyArg(args[argsIndex]);
        argsIndex += 1;
      }
    } else {
      output += str;
    }
  }

  return output;
});
