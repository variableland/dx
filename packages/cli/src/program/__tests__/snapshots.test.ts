import { createTestProgram, execCli, parseProgram } from "test/helpers";
import { afterAll, beforeAll, expect, test, vi } from "vitest";
import { $ } from "~/shell";

const { program } = createTestProgram();

const rootCommands = ["help", "--help", "-v", "--version"];

beforeAll(() => {
  vi.stubEnv("CI", "false");
});

afterAll(() => {
  vi.unstubAllEnvs();
});

for (const cmd of rootCommands) {
  test(`should match ${cmd}`, async () => {
    const { stdout } = await execCli(cmd);

    expect(stdout).toMatchSnapshot();
  });
}

for (const command of program.commands) {
  const cmd = command.name();

  test(`should match help message for ${cmd}`, async () => {
    const { stdout } = await execCli(`${cmd} --help`);

    expect(stdout).toMatchSnapshot();
  });

  test(`should match ${cmd} command`, async () => {
    await parseProgram([cmd]);

    expect(vi.mocked($)).toBeCalledTimes(1);
    expect(vi.mocked($).mock.results[0]?.value).toMatchSnapshot();
  });
}
