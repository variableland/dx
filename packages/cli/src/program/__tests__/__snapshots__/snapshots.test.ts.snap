// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match --help 1`] = `
"🦊 R U N - R U N: The CLI toolbox to fullstack common scripts in Variable Land 👊

Usage: rr|run-run [options] [command]

Options:
  -v, --version          output the version number
  -h, --help             display help for command

Commands:
  format|fmt [options]   format the code 🎨
  lint [options]         lint the code 🧹
  test:static [options]  check format and lint issues ✅
  clean [options]        delete dirty folders or files such as node_modules, etc
                         🗑️
  typecheck|tsc          check if TypeScript code is well typed 🎨
  help [command]         display help for command

Acknowledgment:
- kcd-scripts: for main inspiration
  https://github.com/kentcdodds/kcd-scripts

- peruvian news: in honor to Run Run
  https://es.wikipedia.org/wiki/Run_Run
"
`;

exports[`should match --version 1`] = `
"0.1.0
"
`;

exports[`should match -v 1`] = `
"0.1.0
"
`;

exports[`should match clean command 1`] = `"npx --yes rimraf -g **/.turbo **/dist **/node_modules pnpm-lock.yaml"`;

exports[`should match format command 1`] = `"biome format --no-errors-on-unmatched --colors=force"`;

exports[`should match help 1`] = `
"🦊 R U N - R U N: The CLI toolbox to fullstack common scripts in Variable Land 👊

Usage: rr|run-run [options] [command]

Options:
  -v, --version          output the version number
  -h, --help             display help for command

Commands:
  format|fmt [options]   format the code 🎨
  lint [options]         lint the code 🧹
  test:static [options]  check format and lint issues ✅
  clean [options]        delete dirty folders or files such as node_modules, etc
                         🗑️
  typecheck|tsc          check if TypeScript code is well typed 🎨
  help [command]         display help for command

Acknowledgment:
- kcd-scripts: for main inspiration
  https://github.com/kentcdodds/kcd-scripts

- peruvian news: in honor to Run Run
  https://es.wikipedia.org/wiki/Run_Run
"
`;

exports[`should match help message for clean 1`] = `
"Usage: rr clean [options]

delete dirty folders or files such as node_modules, etc 🗑️

Options:
  --only-dist  delete 'dist' folders only
  -h, --help   display help for command

Under the hood, this command uses the rimraf CLI to delete dirty folders or files.
"
`;

exports[`should match help message for format 1`] = `
"Usage: rr format|fmt [options]

format the code 🎨

Options:
  -c, --check  check if the code is formatted (default: true)
  -f, --fix    format all the code
  -h, --help   display help for command

Under the hood, this command uses the biome CLI to format the code.
"
`;

exports[`should match help message for lint 1`] = `
"Usage: rr lint [options]

lint the code 🧹

Options:
  -c, --check  check if the code is valid (default: true)
  -f, --fix    try to fix all the code
  -h, --help   display help for command

Under the hood, this command uses the biome CLI to lint the code.
"
`;

exports[`should match help message for test:static 1`] = `
"Usage: rr test:static [options]

check format and lint issues ✅

Options:
  -f, --fix     try to fix issues automatically
  --fix-staged  try to fix staged files only
  -h, --help    display help for command

Under the hood, this command uses the biome CLI to check the code.
"
`;

exports[`should match help message for typecheck 1`] = `
"Usage: rr typecheck|tsc [options]

check if TypeScript code is well typed 🎨

Options:
  -h, --help  display help for command

Under the hood, this command uses the TSC CLI to check the code.
"
`;

exports[`should match lint command 1`] = `"biome check --colors=force --formatter-enabled=false"`;

exports[`should match test:static command 1`] = `"biome check --colors=force"`;

exports[`should match typecheck command 1`] = `"tsc --noEmit"`;
